src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs

```diff
@@ -594,34 +587,28 @@ protected virtual void OnCustomCommand(int command)
         ///       contains multiple associated services. Loads the specified services into memory so they can be
         ///       started.</para>
         /// </devdoc>
-        public static void Run(ServiceBase[] services)
+        public static unsafe void Run(ServiceBase[] services)
         {
             if (services == null || services.Length == 0)
                 throw new ArgumentException(SR.NoServices);
 
-            int sizeOfSERVICE_TABLE_ENTRY = Marshal.SizeOf<SERVICE_TABLE_ENTRY>();
-
-            IntPtr entriesPointer = Marshal.AllocHGlobal(checked((services.Length + 1) * sizeOfSERVICE_TABLE_ENTRY));
+            IntPtr entriesPointer = Marshal.AllocHGlobal(checked((services.Length + 1) * Marshal.SizeOf<SERVICE_TABLE_ENTRY>()));
+            Span<SERVICE_TABLE_ENTRY> entries = new Span<SERVICE_TABLE_ENTRY>((void*)entriesPointer, services.Length + 1);
             try
             {
-                SERVICE_TABLE_ENTRY[] entries = new SERVICE_TABLE_ENTRY[services.Length];
                 bool multipleServices = services.Length > 1;
-                IntPtr structPtr;
 
-                for (int index = 0; index < services.Length; ++index)
+                for (int index = 0; index < entries.Length - 1; ++index)
                 {
-                    services[index].Initialize(multipleServices);
-                    entries[index] = services[index].GetEntry();
-                    structPtr = entriesPointer + sizeOfSERVICE_TABLE_ENTRY * index;
-                    Marshal.StructureToPtr(entries[index], structPtr, fDeleteOld: false);
+                    entries[index] = default; // Clear out the entry before we do anything, so if anything fails we can know how much we've successfully allocated so far.
```

Comment:
If something fails half-way through the loop, the half of the entries array is still going to contain garbage.

Instead of this, you should call `entries.Clear()` before the try.

Reply:
(You won't even need to set the sentinel entry to zeros if you do that.)

Reply:
I believe this should be handled by https://github.com/dotnet/corefx/pull/42472/files/39b40d556dc670f4f4a64ba46ba9cc090fe9165c#diff-0c15cbba516668d4903c3399a546c570R653, where I stop freeing memories once I reach a null entry.

calling `entries.Clear` instead of doing this element-by-element is a fair opinion as well, though. The whole of the array will be cleared eventually anyway.

Reply:
Ah, you are right - I have missed this detail in the review. I would either call `entries.Clear()` upfront, or had explicit counter for how many entries needs to be freed (that is always incremented after the entry is assigned). The explicit counter is a bit more efficient, but the efficiency does not really matter in this case.